zerver/views/auth.py

```diff
@@ -312,6 +326,16 @@ def login_or_register_remote_user(request: HttpRequest, result: ExternalAuthResu
     redirect_to = get_safe_redirect_to(redirect_to, user_profile.realm.uri)
     return HttpResponseRedirect(redirect_to)
 
+def handle_action_key(request: HttpRequest, user_profile: UserProfile,
+                      key: str) -> HttpResponse:
+    try:
+        confirmation_object = get_confirmation_object(key)
+    except Confirmation.DoesNotExist:
+        return render(request, 'confirmation/link_does_not_exist.html')
```

Comment:
this block is basically duplicating what happens inside confirm_email_change I think? We should clean this up somehow.

Reply:
Not exactly. In `confirm_email_change`, we knew what `type` the confirmation object was so we were using a very similar function `get_object_from_key` which requires us to pass the object type which in `confirm_email_change`'s case was `EMAIL_CHANGE` but here we cannot know what type of object is stored in the key before we get it once we start using action key for other purposes. So, I added this function just for the sake of knowing what type the object is.

I think we should instead define a `get_confirmation_object_type`. will change it so that the code will be more readable.

Reply:
Can't we simply solve this weird duplicate fetching and validation of the key in both function by making `confirm_email_change` take the already fetched `Confirmation` object as argument rather than the key string?

Reply:
That'd be nice but the existing `get_object_from_key` requires `confirmation_type` to be sent to get the object. If we change it to not require `confirmation_type`, we need to change it in 9 places(excluding settings) it's used. We can define another similar function but it'd be same as the existing `get_object_from_key`.

What would be a better thing to do here? I'm inclined to changing `get_object_from_key` requiring `confirmation_type` but not sure if there would be an issue doing that.

Reply:
As discussed, I changed `handle_action_key` to just get the object specifying the confirmation key type to be email change as it's the only type of action key this flow supports since we can extend the `handle_action_key` later when we add other types of confirmation keys to be `action_key`.